test/integrationtests/voight_kampff/tools.py

```diff
@@ -15,15 +15,188 @@
+        match_event: mechanism for knowing when a match is found
+        error_message: message that can be used by the test to communicate
+            the reason for a failed match to the tester.
+    """
+    def __init__(self, message_type: str, context: Any):
+        self.message_type = message_type
+        self.context = context
+        self.match_event = Event()
+        self.error_message = ""
+
+    @property
+    def match_found(self):
+        return self.match_event.is_set()
+
+    def match(self, timeout: int = None):
+        """Attempts to match the requested message type to emitted bus events.
+
+        Use a message bus event handler to capture any message emitted on the
+        bus that matches the message type specified by the caller.  Also
+        checks any messages emitted prior to the handler being defined to
+        protect against a race condition.
+
+        Args:
+            timeout: number of seconds to attempt matching before giving up
+        """
+        timeout = timeout or self.context.step_timeout
+        self.context.bus.on(self.message_type, self.handle_message)
+        self._check_historical_messages()
+        if not self.match_event.is_set():
+            self.match_event.wait(timeout=timeout)
+        self.context.bus.remove(self.message_type, self.handle_message)
+        if not self.match_found:
+            self._build_error_message()
+
+        return self.match_found, self.error_message
+
+    def _check_historical_messages(self):
+        """Searches messages emitted before the event handler was defined."""
+        for message in self.context.bus.get_messages(self.message_type):
+            self.handle_message(message)
+            if self.match_found:
+                break
+        self.context.bus.clear_messages()
+
+    def handle_message(self, message: Message):
+        """Applies matching criteria to the emitted event.
+
+        Args:
+            message: message emitted by bus with the requested message type
+        """
+        self.context.matched_message = message
+        self.match_event.set()
+
+    def _build_error_message(self):
+        """Builds a message that communicates the failure to the test."""
+        self.error_message = (
+            f"Expected message type {self.message_type} was not emitted."
+        )
+
+
+class VoightKampffDialogMatcher(VoightKampffMessageMatcher):
+    """Variation of VoightKampffEventMatcher for matching dialogs.
+
+    Usage:
+        Intended for use in a single test condition.
+
+        matcher = VoightKampffDialogMatcher(context, dialogs)
+        match_found, error_message = matcher.match
+        assert match_found, error_message
+
+    Attributes:
+        dialogs: one or more dialog names that will constitute a match
+        speak_messages: bus messages with message type of "speak" captured
+            in the matching process
+    """
+    def __init__(self, context: Any, dialogs: List[str]):
```

Comment:
Small nitpick that the args order for the Matcher classes aren't 100% consistent.

Is this an accurate characterization?
```python
VoightKampffMessageMatcher(thing_to_look_for, context)
VoightKampffDialogMatcher(context, thing_to_look_for)
VoightKampffCriteriaMatcher(thing_to_look_for, context, how_to_look)
```

Reply:
Fixed